// music-metadata/lib/riff/RiffParser.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var strtok3 = require("strtok3");
var Token = require("token-types");
var RiffChunk = require("./RiffChunk");
var WaveChunk = require("./../wav/WaveChunk");
var stream_1 = require("stream");
var ID3v2Parser_1 = require("../id3v2/ID3v2Parser");
var Common_1 = require("../common/Common");
var FourCC_1 = require("../common/FourCC");
/**
 * Resource Interchange File Format (RIFF) Parser
 *
 * WAVE PCM soundfile format
 *
 * Ref:
 *  http://www.johnloomis.org/cpe102/asgn/asgn1/riff.html
 *  http://soundfile.sapp.org/doc/WaveFormat
 */
var WavePcmParser = /** @class */ (function () {
    function WavePcmParser() {
        this.metadata = {
            format: {
                dataformat: "WAVE PCM"
            },
            native: {}
        };
        /**
         * RIFF/ILIST-INFO tag stored in EXIF
         */
        this.riffInfoTags = [];
        this.warnings = [];
    }
    WavePcmParser.prototype.parse = function (tokenizer, options) {
        var _this = this;
        this.tokenizer = tokenizer;
        this.options = options;
        return this.tokenizer.readToken(RiffChunk.Header)
            .then(function (header) {
            if (header.chunkID !== 'RIFF')
                return null; // Not RIFF format
            return _this.tokenizer.readToken(FourCC_1.FourCcToken).then(function (type) {
                _this.metadata.format.dataformat = type;
            }).then(function () {
                return _this.readChunk(header).then(function () {
                    return null;
                });
            });
        })
            .catch(function (err) {
            if (err.message === strtok3.endOfFile) {
                return _this.metadata;
            }
            else {
                throw err;
            }
        }).then(function (metadata) {
            if (_this.riffInfoTags.length > 0) {
                metadata.native.exif = _this.riffInfoTags;
            }
            return metadata;
        });
    };
    WavePcmParser.prototype.readChunk = function (parent) {
        var _this = this;
        return this.tokenizer.readToken(WaveChunk.Header)
            .then(function (header) {
            switch (header.chunkID) {
                case "LIST":
                    return _this.tokenizer.readToken(FourCC_1.FourCcToken).then(function (listTypes) {
                        switch (listTypes) {
                            case 'INFO':
                                return _this.parseRiffInfoTags(header.size - 4).then(function () { return header.size; });
                            default:
                                _this.warnings.push("Ignore chunk: RIFF/LIST." + listTypes);
                                return _this.tokenizer.ignore(header.size).then(function () { return header.size; });
                        }
                    });
                case "fmt ":// The Common Chunk
                    return _this.tokenizer.readToken(new WaveChunk.Format(header))
                        .then(function (common) {
                        _this.metadata.format.bitsPerSample = common.bitsPerSample;
                        _this.metadata.format.sampleRate = common.sampleRate;
                        _this.metadata.format.numberOfChannels = common.numChannels;
                        _this.metadata.format.bitrate = common.blockAlign * common.sampleRate * 8;
                        _this.blockAlign = common.blockAlign;
                    });
                case "id3 ": // The way Picard, FooBar currently stores, ID3 meta-data
                case "ID3 ":// The way Mp3Tags stores ID3 meta-data
                    return _this.tokenizer.readToken(new Token.BufferType(header.size))
                        .then(function (id3_data) {
                        var id3stream = new ID3Stream(id3_data);
                        return strtok3.fromStream(id3stream).then(function (rst) {
                            return ID3v2Parser_1.ID3v2Parser.getInstance().parse(_this.metadata, rst, _this.options);
                        });
                    });
                case 'data':// PCM-data
                    _this.metadata.format.numberOfSamples = header.size / _this.blockAlign;
                    _this.metadata.format.duration = _this.metadata.format.numberOfSamples / _this.metadata.format.sampleRate;
                    _this.metadata.format.bitrate = _this.metadata.format.numberOfChannels * _this.blockAlign * _this.metadata.format.sampleRate; // ToDo: check me
                    return _this.tokenizer.ignore(header.size);
                default:
                    _this.warnings.push("Ignore chunk: RIFF/" + header.chunkID);
                    return _this.tokenizer.ignore(header.size);
            }
        }).then(function () {
            return _this.readChunk(parent);
        });
    };
    WavePcmParser.prototype.parseRiffInfoTags = function (chunkSize) {
        var _this = this;
        return this.tokenizer.readToken(WaveChunk.Header)
            .then(function (header) {
            var valueToken = new RiffChunk.ListInfoTagValue(header);
            return _this.tokenizer.readToken(valueToken).then(function (value) {
                _this.riffInfoTags.push({ id: header.chunkID, value: Common_1.default.stripNulls(value) });
                chunkSize -= (8 + valueToken.len);
                if (chunkSize > 8) {
                    return _this.parseRiffInfoTags(chunkSize);
                }
                else if (chunkSize === 0) {
                    return Promise.resolve();
                }
                else {
                    throw Error("Illegal remaining size: " + chunkSize);
                }
            });
        });
    };
    return WavePcmParser;
}());
exports.WavePcmParser = WavePcmParser;
var ID3Stream = /** @class */ (function (_super) {
    __extends(ID3Stream, _super);
    function ID3Stream(buf) {
        var _this = _super.call(this) || this;
        _this.buf = buf;
        return _this;
    }
    ID3Stream.prototype._read = function () {
        this.push(this.buf);
        this.push(null); // push the EOF-signaling `null` chunk
    };
    return ID3Stream;
}(stream_1.Readable));
